### PR TITLE
chore(ui): migrate JsonViewer to typescript

### PR DESCRIPTION
**Pull request by gjaskiewicz**
# Summary

Migrate JsonViewer to typescript

# Changes Made

- migrate JsonViewer to typescript

# Related Issues

- Issue 1: [https://github.com/cloudoperators/juno/issues/232]

# Testing Instructions

1. `npm i`
2. `npm run storybook`

# Checklist

- [ ] I have performed a self-review of my code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have made corresponding changes to the documentation (if applicable).
- [ ] My changes generate no new warnings or errors.

### CODE REVIEW

File: packages/ui-components/src/components/JsonViewer/JsonViewer.component.tsx [hunk 1]
```diff
@@ -0,0 +1,607 @@
...
+    nan?: string
+    null?: string
+    undefined?: string
+    regexp?: string
+    background?: string
+  }
+}
+
+interface ThemeContextType {
+  //extends themes.JsonViewerTheme {
```
Comment (andypf) [inline packages/ui-components/src/components/JsonViewer/JsonViewer.component.tsx:130]: we can remove this line I guess

**Review by andypf — COMMENTED** [review]
just one small comment from me. 

I can’t say much about TypeScript, but the scope is definitely large, and for that: good job!

**Review by edda — CHANGES_REQUESTED** [review]
The storybook preview is broken. If I read the error message correctly it's because the deprecated version still contains stories:
```
Error: Unable to index files:
- ./src/components/JsonViewer/JsonViewer.stories.tsx,./src/deprecated_js/JsonViewer/JsonViewer.stories.js: Duplicate stories with id: components-jsonviewer--default
- ./src/components/JsonViewer/JsonViewer.stories.tsx,./src/deprecated_js/JsonViewer/JsonViewer.stories.js: Duplicate stories with id: components-jsonviewer--array-data
- ./src/components/JsonViewer/JsonViewer.stories.tsx,./src/deprecated_js/JsonViewer/JsonViewer.stories.js: Duplicate stories with id: components-jsonviewer--light
- ./src/components/JsonViewer/JsonViewer.stories.tsx,./src/deprecated_js/JsonViewer/JsonViewer.stories.js: Duplicate stories with id: components-jsonviewer--expanded
- ./src/components/JsonViewer/JsonViewer.stories.tsx,./src/deprecated_js/JsonViewer/JsonViewer.stories.js: Duplicate stories with id: components-jsonviewer--with-toolbar
- ./src/components/JsonViewer/JsonViewer.stories.tsx,./src/deprecated_js/JsonViewer/JsonViewer.stories.js: Duplicate stories with id: components-jsonviewer--custom-theme
```

File: packages/ui-components/src/components/JsonViewer/JsonViewer.component.tsx [hunk 2]
```diff
@@ -0,0 +1,606 @@
...
+  if (value instanceof RegExp) return "regexp"
+  if (value instanceof Date) return "date"
+  const t = (typeof value).toLowerCase()
+  if (t === "number") {
+    if (Number.isNaN(value)) return "nan"
+    return Number.isInteger(value) ? "integer" : "float"
+  } else return t as TypeValueLabelType
+}
+
+interface ThemeColors {
```
Comment (barsukov) [inline packages/ui-components/src/components/JsonViewer/JsonViewer.component.tsx:90]:
```

type HighlightColors = {
  foreground?: string
  background?: string
}

type ToolbarColors = {
  border?: string
} & HighlightColors

type IconColors = {
  expanded?: string
  collapsed?: string
  expandAll?: string
}

type DataTypeColors = {
  array?: string
  object?: string
  boolean?: string
  date?: string
  float?: string
  function?: string
  integer?: string
  string?: string
  nan?: string
  null?: string
  undefined?: string
  regexp?: string
  background?: string
}

interface ThemeColors {
  background?: string
  ellipsis?: string
  brace?: string
  key?: string
  index?: string
  size?: string
  border?: string
  highlight: HighlightColors
  toolbar: ToolbarColors
  icon: IconColors
  dataType: DataTypeColors
}

```

Comment (barsukov) [reply packages/ui-components/src/components/JsonViewer/JsonViewer.component.tsx:90]: I would modify it this way to split the big type on chunks and also remove some duplciations that way

File: packages/ui-components/src/components/JsonViewer/JsonViewer.component.tsx [hunk 3]
```diff
@@ -0,0 +1,606 @@
...
+        <path d="M1344 800v64q0 14-9 23t-23 9h-832q-14 0-23-9t-9-23v-64q0-14 9-23t23-9h832q14 0 23 9t9 23zm128 448v-832q0-66-47-113t-113-47h-832q-66 0-113 47t-47 113v832q0 66 47 113t113 47h832q66 0 113-47t47-113zm128-832v832q0 119-84.5 203.5t-203.5 84.5h-832q-119 0-203.5-84.5t-84.5-203.5v-832q0-119 84.5-203.5t203.5-84.5h832q119 0 203.5 84.5t84.5 203.5z"></path>
+      ) : (
+        <path d="M1344 800v64q0 14-9 23t-23 9h-352v352q0 14-9 23t-23 9h-64q-14 0-23-9t-9-23v-352h-352q-14 0-23-9t-9-23v-64q0-14 9-23t23-9h352v-352q0-14 9-23t23-9h64q14 0 23 9t9 23v352h352q14 0 23 9t9 23zm128 448v-832q0-66-47-113t-113-47h-832q-66 0-113 47t-47 113v832q0 66 47 113t113 47h832q66 0 113-47t47-113zm128-832v832q0 119-84.5 203.5t-203.5 84.5h-832q-119 0-203.5-84.5t-84.5-203.5v-832q0-119 84.5-203.5t203.5-84.5h832q119 0 203.5 84.5t84.5 203.5z"></path>
+      )}
+    </svg>
+  )
+}
+
+// Toolbar Icon
+const ExpandAllIcon = () => {
```
Comment (barsukov) [inline packages/ui-components/src/components/JsonViewer/JsonViewer.component.tsx:189]: I would extract this one and other one to nearby components but exported them only to make use inside of this component

File: packages/ui-components/src/components/JsonViewer/JsonViewer.component.tsx [hunk 4]
```diff
@@ -0,0 +1,606 @@
...
+    string?: string
+    nan?: string
+    null?: string
+    undefined?: string
+    regexp?: string
+    background?: string
+  }
+}
+
+interface ThemeContextType {
```
Comment (barsukov) [inline packages/ui-components/src/components/JsonViewer/JsonViewer.component.tsx:129]: I would extract almost all types to nearby file and called it "JsonViewerTypes.types.ts" but without PropTypes interfaces that should be belong to components.

File: packages/ui-components/src/components/JsonViewer/JsonViewer.component.tsx [hunk 5]
```diff
@@ -0,0 +1,606 @@
...
+    nan: theme.base08,
+    null: theme.base0A,
+    undefined: theme.base05,
+    regexp: theme.base0A,
+    background: theme.base02,
+  },
+})
+
+// get type of value
+const type = (value: unknown): TypeValueLabelType => {
```
Comment (barsukov) [inline packages/ui-components/src/components/JsonViewer/JsonViewer.component.tsx:78]: this method has also potential to refactor with a new typescript way, first way rename it from type to maybe getValueTypeLabel and remove the comment, second:

one of my attemption:
```

function getValueTypeFromObject(value: object | null) {
  if (value === null) return "null"
  if (Array.isArray(value)) return "array"
  if (value instanceof RegExp) return "regexp"
  if (value instanceof Date) return "date"
  return "object"
}

function getValueTypeFromNumber(value: number | null) {
  if (Number.isNaN(value)) return "nan"
  return Number.isInteger(value) ? "integer" : "float"
}

const getTypeOfTheValue = (value: unknown): TypeValueLabelType => {
  switch (typeof value) {
    case "object":
      getValueTypeFromObject(value)
    case "number":
      getValueTypeFromNumber(value as number)
    case "boolean":
      return "boolean"
    case "string":
      return "string"
    case "undefined":
      return "undefined"
    case "function":
      return "function"
    default:
      return "background"
  }
}
```

**Review by barsukov — CHANGES_REQUESTED** [review]
My few cents for improving this component

File: packages/ui-components/package.json [hunk 6]
```diff
@@ -101,7 +101,7 @@
     "react-dom": "^18.2.0"
   },
   "scripts": {
-    "storybook": "storybook dev -p $APP_PORT --ci --no-open",
+    "storybook": "storybook dev -p 7301 --ci --no-open",
```
Comment (barsukov) [inline packages/ui-components/package.json:104]: please revert this changes

**Review by barsukov — CHANGES_REQUESTED** [review]
Very important to not forget to rollback changes and fix other errors

**Review by barsukov — CHANGES_REQUESTED** [review]
index.js should be updated too please pay attention to it, do not duplicate parts just remove the .js extensions.

**Review by barsukov — DISMISSED** [review]
LGTM!

**Review by andypf — APPROVED** [review]
lgtm

**Review by edda — CHANGES_REQUESTED** [review]
The storybook documentation for this component looks very ugly because of the superlong strings in the test data. Please remove those edge cases from the test data so that the documentation is better readable.
Screenshot 2024-10-25 at 13 55 13

**Review by edda — APPROVED** [review]

**Review by barsukov — APPROVED** [review]

**Review by andypf — APPROVED** [review]
lgtm
